### PR TITLE
add azure key vault secrets lookup plugin

### PR DESCRIPTION
**Pull request by yungezz**
add azure key vault secerts lookup plugin in role. @caohai

### CODE REVIEW

File: lookup_plugins/azure_keyvault_secret.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        - requests
+        - azure
+        - msrest
+    short_description: read secret from Azure Key Vault.
+    description:
+      - This lookup returns the content of a secret kept in Azure Key Vault.
+    options:
+        _terms:
+            description: secret name of the secret to retrieve, version can be included like secret_name/secret_version.
+            required: True
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: maybe all descriptions should start from uppercase and just put them into new line wit "-" for consistenct

File: lookup_plugins/azure_keyvault_secret.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        client_id:
+            description: client_id of service principal that has access to the provided Azure Key Vault
+        secret:
+            description: secret of service principal provided above
+        tenant_id:
+            description: tenant_id of service principal provided above
+    notes:
+        - If version is not provided, Key Vault will give the latest version.
+        - If ansible is running on Azure Virtual Machine with MSI enabled, client_id, secret and tenant isn't necessary.
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: I would change "necessary" to "required".

File: lookup_plugins/azure_keyvault_secret.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        - For how to enable MSI on Azure VM, please refer to this doc https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: "For how to enable" could be change to "For enabling", or even better "To enable"

File: lookup_plugins/azure_keyvault_secret.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        - After enableing MSI on Azure VM, remember to grant this VM access to the Key Vault by adding a new Acess Policy in Azure Portal.
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: enableing -> enabling

and also 

"to grant access to the Key Vault to the VM" perhaps?

File: lookup_plugins/azure_keyvault_secret.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        - If MSI is not available on the machine, then you have to provide a valid service principal that has access to the key vault.
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: "then you have" doesnt' sound good. Instead:


"If MSI is not available on the VM, it's necessary to provide......"

File: lookup_plugins/azure_keyvault_secret.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+"""
+
+EXAMPLE = """
+- name: Lookup secret via MSI endpoint
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: Lookup -> Look up

File: lookup_plugins/azure_keyvault_secret.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+- name: Lookup secret via KeyVault Client
+  vars:
+    url: 'https://yourvault.vault.azure.net'
+    secretname: 'testSecret/version'
+    client_id: '123456789'
+    secret: 'abcdefg'
+    tenant: 'uvwxyz'
+  debug: msg="the value of this secret is {{lookup('azure_keyvault_secret',secretname,vault_url=url, cliend_id=client_id, secret=secret, tenant_id=tenant)}}"
+
+# Example below creates an Azure Virtual Machine with ssh public key from key vault using 'azure_keyvault_secret' lookup plugin.
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: ssh -> SSH

File: lookup_plugins/azure_keyvault_secret.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+      image:
+        offer: UbuntuServer
+        publisher: Canonical
+        sku: 16.04-LTS
+        version: latest
+"""
+
+RETURN = """
+  _raw:
+    description: secret content
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: what is that content exactly? is it string?

File: lookup_plugins/azure_keyvault_secret.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+try:
+    token_res = requests.get('http://169.254.169.254/metadata/identity/oauth2/token', params=token_params, headers=token_headers)
+    token = token_res.json()["access_token"]
+    TOKEN_ACQUIRED = True
+except requests.exceptions.RequestException:
+    print('Unable to fetch MSI token. Will use service principal if provided.')
+    TOKEN_ACQUIRED = False
+
+
+def lookup_sercret_non_msi(terms, vault_url, kwargs):
```
Comment (zikalino) [inline lookup_plugins/azure_keyvault_secret.py]: sercret -> secret

**Comment by yungezz** [comment]
@zikalino , thanks for review, comments resolved. pls help to review again.

**Comment by Fred-sun** [comment]
@zikalino @yuwzho comments resolved , Please help to review again! Thanks!

File: lookup_plugins/azure_keyvault_secret.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    secret = kwargs.pop('secret', None)
+    tenant_id = kwargs.pop('tenant_id', None)
+
+    try:
+        credentials = ServicePrincipalCredentials(
+            client_id=client_id,
+            secret=secret,
+            tenant=tenant_id
+        )
+        client = KeyVaultClient(credentials)
```
Comment (yuwzho) [inline lookup_plugins/azure_keyvault_secret.py]: Need we track UA here?

File: lookup_plugins/azure_keyvault_secret.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    TOKEN_ACQUIRED = False
+
+
+def lookup_secret_non_msi(terms, vault_url, kwargs):
+    import logging
+    logging.getLogger('msrestazure.azure_active_directory').addHandler(logging.NullHandler())
+    logging.getLogger('msrest.service_client').addHandler(logging.NullHandler())
+
+    try:
+        from azure.common.credentials import ServicePrincipalCredentials
```
Comment (yuwzho) [inline lookup_plugins/azure_keyvault_secret.py]: Need specific the lib version?

**Comment by Fred-sun** [comment]
@zikalino Could you help to resolve the issues mentioned in the comments?

**Comment by Fred-sun** [comment]
@yuwzho Could you help review this PR again? Thanks!

**Pull request merged**